presto-main/src/main/java/com/facebook/presto/server/StatementResource.java

```diff
@@ -483,46 +517,81 @@ else if (!queryInfo.getOutputStage().isPresent()) {
             return queryResults;
         }
 
-        private synchronized Iterable<List<Object>> getData(Duration maxWait)
-                throws InterruptedException
+        private CompletableFuture<?> waitForQueryStart(Duration maxWaitTime)
         {
-            // wait for query to start
-            QueryInfo queryInfo = queryManager.getQueryInfo(queryId);
-            while (maxWait.toMillis() > 1 && !isQueryStarted(queryInfo)) {
-                queryManager.recordHeartbeat(queryId);
-                maxWait = queryManager.waitForStateChange(queryId, queryInfo.getState(), maxWait);
-                queryInfo = queryManager.getQueryInfo(queryId);
+            CompletableFuture<?> future = new CompletableFuture<>();
+            QueryState state = queryManager.getQueryInfo(queryId).getState();
+            waitForQueryState(future, state, Query::isQueryStarted);
+            return addTimeout(future, () -> null, maxWaitTime, timeoutExecutor);
+        }
+
+        private CompletableFuture<?> waitForQueryFinish(Duration maxWaitTime)
+        {
+            // check if we have received all of the output data
+            if (!exchangeClient.isClosed()) {
+                return completedFuture(null);
+            }
+
+            // wait for the query to finish
+            CompletableFuture<?> future = new CompletableFuture<>();
+            QueryState state = queryManager.getQueryInfo(queryId).getState();
+            waitForQueryState(future, state, QueryState::isDone);
+            return addTimeout(future, () -> null, maxWaitTime, timeoutExecutor);
+        }
+
+        private void waitForQueryState(CompletableFuture<?> future, QueryState currentState, Predicate<QueryState> predicate)
+        {
+            if (predicate.test(currentState)) {
+                future.complete(null);
+            }
+            else {
+                queryManager.getStateChange(queryId, currentState)
+                        .whenComplete((state, throwable) -> waitForQueryState(future, state, predicate));
```

Comment:
This method is called asynchronously from the future callback, so it's not actually recursive. (unless I'm missing something ... I have a hard time thinking about async callbacks)